typeinf.ml

```diff
@@ -0,0 +1,68 @@
+open Lang
+
+
+exception Error of string 
+
+
+(* Constraints *)
+type constr = (tml_type * tml_type) list
+
+let merge_constr : constr -> constr -> constr = fun c1 c2 -> c1 @ c2
+
+type ty_var = string
+
+let genFresh : unit -> string =
+  let counter = ref 0 in
+  fun _ -> counter := !counter + 1;
+           "?X_" ^ (string_of_int (!counter-1))
+
+
+(* Type environment *)
+type environment_type = (string * tml_type) list
+
+let extend_env k v (env: environment_type) = (k, v) :: env
+
+let rec lookup (var: string) (env: environment_type): tml_type option =
+  match env with
+  | [] -> None
+  | (variable, value) :: env' -> if variable = var
+                                 then Some value
+                                 else lookup var env'
+
+(* Generate a list of constraints and a type *)
+let rec infereConstraints_exp (global_gamma : environment_type) (gamma : tml_type list) : tml_expr -> constr * tml_type = function
```

Comment:
How does constraint inference connect to type inference? Can you provide a single function `infer_types` that would return a map for types of all subexpressions?

Also please rename to avoid camlCase, i.e. into infer_constraints_exp.